src/Nethermind/Nethermind.Evm/EvmObjectFormat.cs

```diff
@@ -0,0 +1,236 @@
+            => _logger = logger;
+
+        // magic prefix : EofFormatByte is the first byte, EofFormatDiff is chosen to diff from previously rejected contract according to EIP3541
+        private const byte EofMagicLength = 2;
+        private const byte EofFormatByte = 0xEF;
+        private const byte EofFormatDiff = 0x00;
+        private byte[] EofMagic => new byte[] { EofFormatByte, EofFormatDiff };
+
+        public bool HasEOFFormat(ReadOnlySpan<byte> code) => code.Length > EofMagicLength && code.StartsWith(EofMagic);
+        public bool ExtractHeader(ReadOnlySpan<byte> code, IReleaseSpec spec, out EofHeader header)
+        {
+            if (!HasEOFFormat(code))
+            {
+                if (LoggingEnabled)
+                {
+                    _logger.Trace($"EIP-3540 : Code doesn't start with Magic byte sequence expected {EofMagic.ToHexString(true)} ");
+                }
+                header = null; return false;
+            }
+
+            int codeLen = code.Length;
+
+            int i = EofMagicLength;
+            byte EOFVersion = code[i++];
+
+            header = new EofHeader
+            {
+                Version = EOFVersion
+            };
+
+            switch (EOFVersion)
+            {
+                case 1:
+                    return HandleEOF1(spec, code, ref header, codeLen, ref i);
+                default:
+                    if (LoggingEnabled)
+                    {
+                        _logger.Trace($"EIP-3540 : Code has wrong EOFn version expected {1} but found {EOFVersion}");
+                    }
+                    header = null; return false;
+            }
+        }
+
+        private bool HandleEOF1(IReleaseSpec spec, ReadOnlySpan<byte> code, ref EofHeader header, int codeLen, ref int i)
+        {
+            bool continueParsing = true;
+
+            UInt16? CodeSections = null;
+            UInt16? DataSections = null;
+
+            while (i < codeLen && continueParsing)
+            {
+                var sectionKind = (SectionDividor)code[i];
+                i++;
+
+                switch (sectionKind)
+                {
+                    case SectionDividor.Terminator:
+                        {
+                            if (CodeSections is null || CodeSections == 0)
+                            {
+                                if (LoggingEnabled)
+                                {
+                                    _logger.Trace($"EIP-3540 : No Codesection found");
+                                }
+                                header = null; return false;
+                            }
+
+                            header.CodeSize = CodeSections.Value;
+                            header.DataSize = DataSections ?? 0;
+                            continueParsing = false;
+                            break;
+                        }
+                    case SectionDividor.CodeSection:
+                        {
```

Comment:
Lets add check to make sure CodeSection is null to solve the case of having 2 code sections after each other.